source/common/orca/orca_parser.cc

```diff
@@ -0,0 +1,178 @@
+    auto metric_name_without_prefix = absl::StripPrefix(metric_name, kNamedMetricsFieldPrefix);
+    return tryCopyNamedMetricToOrcaLoadReport(metric_name_without_prefix, value, orca_load_report);
+  }
+
+  if (metric_name == kCpuUtilizationField) {
+    orca_load_report.set_cpu_utilization(value);
+  } else if (metric_name == kMemUtilizationField) {
+    orca_load_report.set_mem_utilization(value);
+  } else if (metric_name == kApplicationUtilizationField) {
+    orca_load_report.set_application_utilization(value);
+  } else if (metric_name == kEpsField) {
+    orca_load_report.set_eps(value);
+  } else if (metric_name == kRpsFractionalField) {
+    orca_load_report.set_rps_fractional(value);
+  } else {
+    return absl::InvalidArgumentError(absl::StrCat("unsupported metric name: ", metric_name));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status tryParseNativeHttpEncoded(const HeaderMap::GetResult& header,
+                                       OrcaLoadReport& orca_load_report) {
+  const std::vector<absl::string_view> values = parseCommaDelimitedHeader(header);
+
+  // Check for duplicate metric names here because OrcaLoadReport fields are not
+  // marked as optional and therefore don't differentiate between unset and
+  // default values.
+  absl::flat_hash_set<absl::string_view> metric_names;
+  for (const auto value : values) {
+    std::pair<absl::string_view, absl::string_view> entry =
+        absl::StrSplit(value, absl::MaxSplits(':', 1), absl::SkipWhitespace());
+    if (metric_names.contains(entry.first)) {
+      return absl::AlreadyExistsError(
+          absl::StrCat(kEndpointLoadMetricsHeader, " contains duplicate metric: ", entry.first));
+    }
+    RETURN_IF_NOT_OK(tryCopyMetricToOrcaLoadReport(entry.first, entry.second, orca_load_report));
+    metric_names.insert(entry.first);
+  }
+  return absl::OkStatus();
+}
+
+} // namespace
+
+absl::StatusOr<OrcaLoadReport> parseOrcaLoadReportHeaders(const HeaderMap& headers) {
+  OrcaLoadReport load_report;
+
+  const auto load_metrics_native_http =
+      headers.get(Envoy::Http::LowerCaseString(kEndpointLoadMetricsHeader));
+  const auto load_metrics_json =
+      headers.get(Envoy::Http::LowerCaseString(kEndpointLoadMetricsHeaderJson));
+  const auto load_metrics_bin =
+      headers.get(Envoy::Http::LowerCaseString(kEndpointLoadMetricsHeaderBin));
+
+  if (load_metrics_native_http.empty() && load_metrics_json.empty() && load_metrics_bin.empty()) {
+    return absl::NotFoundError("no ORCA data sent from the backend");
+  }
+
+  if ((!load_metrics_native_http.empty() && !load_metrics_json.empty()) ||
+      (!load_metrics_native_http.empty() && !load_metrics_bin.empty()) ||
+      (!load_metrics_json.empty() && !load_metrics_bin.empty())) {
+    // If more than one ORCA header format is found, we will be
+    // unable to determine which header to use.
+    return absl::InvalidArgumentError("more than one ORCA header found");
```

Comment:
I checked the design doc of ORCA, it said `ORCA must support at least one of these to be viable as an open standard.` It seems a load balancer may not implement all three formats. Also, it seems there is no way for a backend to discover which format is supported by load balancer. 

So the question is, is there a case of the backend may send multiple formats since the backend doesn't know which format supported by the load balancer?

If yes, then it seems like we shouldn't reject if there are multiple formats that show up.

Reply:
Rather than three different headers, I think only one header with additional flag to tell which format is used would be better. Like `JSON {...}` `PROTO <base64 encoded binary>`.

And it's will has better performance because we needn't to search 3 headers.

Reply:
Our internal design docs outline that the Envoy load balancer is expected to support all three formats, but enforce that only one is sent from the backend for any given request.

CCing @efimki

Reply:
I believe `ORCA must support at least one of these to be viable as an open standard.` is pointing out minimum support required for this feature. This implementation exceeds the minimum requirements.

Furthermore, configuring header format for the load balancer would mean each load balancer can only support a single format, which is limiting for load balancers proxying requests from backends belonging to outside parties.

Reply:
I've added `Header format` to `ORCA Parser` section of updated the [design doc](https://docs.google.com/document/d/1gb_2pcNnEzTgo1EJ6w1Ol7O-EH-O_Ysu5o215N9MTAg/edit?usp=sharing).

I think I understand the concern about LB rejecting multiple headers without any way for backend to know which header format is supported by particular LB. Maybe we should define the order of priorities instead?

Reply:
> I've added `Header format` to `ORCA Parser` section of the [design doc](https://docs.google.com/document/d/1gb_2pcNnEzTgo1EJ6w1Ol7O-EH-O_Ysu5o215N9MTAg/edit?usp=sharing).
> 
> I think I understand the concern about LB rejecting multiple headers without any way for backend to know which header format is supported by particular LB. Maybe we should define the order of priorities instead?

Yea, I think the order of priorities works. Also, we can let the LB to tell backend which format is supported, but that is something can be think about later.

Reply:
Thank you for the feedback. Updated the PR